refactor(core): Ensure runtime checks rely on roles and scopes

ivov opened the pull request:
https://linear.app/n8n/issue/PAY-1253

ivov committed:
Initial setup

ivov committed:
Make `findProjectIds` distinct

ivov committed:
Cleanup

ivov committed:
Return early if no projects

ivov committed:
Add role service

ivov committed:
Better phrasing

ivov committed:
Add TODO

ivov committed:
Expose project IDs

ivov committed:
Cleanup

ivov committed:
Typo

ivov committed:
Set up `isAccessible`

ivov committed:
Add TODO

Commit:
Merge branch 'feature/rbac' into pay-1253-check-execution-permissions-using-roles-and-scopes

Commit:
Merge remote-tracking branch 'origin/feature/rbac' into pay-1253-check-execution-permissions-using-roles-and-scopes

Commit:
feat: Use project relations for runtime execution credential checks

Commit:
remove unused file

krynble commented on packages/cli/src/UserManagement/PermissionChecker.ts:
```diff
@@ -1,68 +1,76 @@
-		const user = await this.userRepository.findOneOrFail({
-			where: { id: userId },
-		});
+		const user = await this.userRepository.findOneByOrFail({ id: userId });
```
Would it be possible to remove `userId` and `user` entirely from the equation?

Since all executions should be part of saved workflows, the `userId` should not be needed anymore.

One caveat is that we can execute workflows form file or from JSON using the `Execute Workflow` node, but those should inherit the permissions from the parent workflow, so we should still be able to rely on `workflowId` above.

WDYT?

krynble commented on packages/cli/src/UserManagement/PermissionChecker.ts:
```diff
@@ -1,68 +1,76 @@
-
-		// allow if all creds used in this workflow are a subset of
-		// all creds accessible to users who have access to this workflow
+		const { roles, projectIds } = await this.projectService.findRolesAndProjects(
```
Would it make sense to change the scope checks to the middleware instead?

I'm asking because the way it's currently implemented, we might have some unexpected downsides in the future.

Say I'm a regular `Member` in an instance, and create a number of workflows. Then at some point my role is changed to `Viewer` - I would expect to no longer be able to manually execute workflows, but the production workflows I own (i.e. are part of my own personal project) should still continue to run, I think.

This would break with the current code.

krynble replied:
Also I'd say this function should be used to check for credential permissions alone, and not user permissions too. This would be violating the Single Responsibility Principle.

krynble commented on packages/cli/src/databases/repositories/projectRelation.repository.ts:
```diff
@@ -8,6 +8,18 @@ export class ProjectRelationRepository extends Repository<ProjectRelation> {
+			select: ['role'],
+		});
+
+		return new Set(rows.map((row) => row.role));
```
i'd expect this function to return a map of `projectId -> userRole`. Say I'm an admin in 2 projects, viewer in 3 others and a regular member of 5 others, this would return a single Set containing `['admin', 'viewer', 'member']`. I find this confusing.

krynble commented on packages/cli/src/services/project.service.ts:
```diff
@@ -10,15 +10,31 @@ import { type Scope } from '@n8n/permissions';
+	 * Find all the projects where a workflow is accessible,
+	 * along with the roles of a user in those projects.
+	 */
+	async findRolesAndProjects(userId: string, resourceId: string) {
```
```suggestion
	async findRolesAndProjects(userId: string, workflowId: string) {
```
nit

Commit:
remove user ID from PermissionChecker check, replace
`getWorkflowOwnerCached`, and add workflow ID to the manual run endpoint
with permission check

Commit:
update function description

Commit:
Merge remote-tracking branch 'origin/feature/rbac' into pay-1253-check-execution-permissions-using-roles-and-scopes

Commit:
remove now unused function from projectRelation repo
